server/src/main/java/org/opensearch/common/blobstore/stream/read/listener/FileCompletionListener.java

```diff
@@ -0,0 +1,46 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.common.blobstore.stream.read.listener;
+
+import org.opensearch.core.action.ActionListener;
+
+import java.util.concurrent.atomic.AtomicInteger;
+
+/**
+ * FileCompletionListener listens for completion of fetch on all the streams for a file, where
+ * individual streams are handled using {@link StreamCompletionRunnable}. The {@link StreamCompletionRunnable}(s)
+ * hold a reference to the file completion listener to be notified.
+ *
+ * @opensearch.internal
+ */
+public class FileCompletionListener implements ActionListener<Void> {
+    private final int numberOfParts;
+    private final String fileName;
+    private AtomicInteger completedPartsCount;
+    private final ActionListener<String> completionListener;
+
+    public FileCompletionListener(int numberOfParts, String fileName, ActionListener<String> completionListener) {
+        this.completedPartsCount = new AtomicInteger();
+        this.numberOfParts = numberOfParts;
+        this.fileName = fileName;
+        this.completionListener = completionListener;
+    }
+
+    @Override
+    public void onResponse(Void unused) {
```

Comment:
Sorry, I know this was my earlier suggestion, but I don't love the `Void` signature. It seems like communicating the part number is a better API?